heron/schedulers/src/java/com/twitter/heron/scheduler/kubernetes/KubernetesScheduler.java

```diff
@@ -253,4 +370,46 @@ protected String getFetchCommand() {
 
     return command;
   }
+
+  @Override
+  public void addContainers(Set<PackingPlan.ContainerPlan> containersToAdd) {
+    // grab the base pod so we can copy and modify some stuff
+    String basePodName = Runtime.topologyName(runtime) + "-0";
+    JsonNode podConfig;
+    try {
+      podConfig = controller.getBasePod(basePodName);
+    } catch (IOException ioe) {
+      LOG.log(Level.SEVERE, "Problem retrieving base pod configuration", ioe);
+      throw new TopologyRuntimeManagementException("Unable to add containers for update");
+    }
+
+    // iterate over the containers we need to add and rebuild the spec based on the new plan
+    ObjectMapper mapper = new ObjectMapper();
+    for (PackingPlan.ContainerPlan containerPlan : containersToAdd) {
+      String newContainer = rebuildKubernetesPodSpec(podConfig, mapper, containerPlan, 0);
+
+      // deploy this new container
+      try {
+        controller.deployContainer(newContainer);
+      } catch (IOException ioe) {
+        LOG.log(Level.SEVERE, "Problem deploying new container", ioe);
+        throw new TopologyRuntimeManagementException("Unable to add container for update");
+      }
+
+    }
+
+  }
+
+  @Override
+  public void removeContainers(Set<PackingPlan.ContainerPlan> containersToRemove) {
+    for (PackingPlan.ContainerPlan container : containersToRemove) {
+      String podName = Runtime.topologyName(runtime) + "-" + container.getId();
+      try {
+        controller.removeContainer(podName);
+      } catch (IOException ioe) {
+        LOG.log(Level.SEVERE, "Problem removing a container", ioe);
+        throw new TopologyRuntimeManagementException("Unable to remove container for update");
```

Comment:
General comment about logging and exceptions... Typically we either log an error or throw an exception, but not both. Doing both tends to clutters up the logs with duplicate info. Let's instead just throw here, and include more contextual info in the exception message to clarify which container we couldn't remove. Also we should always chain the exceptions so we can get a good stack trace. So something like this:
```
throw new TopologyRuntimeManagementException("Unable to remove container with id " + container.getId(), ioe);
```
Comment applies here and elsewhere in the patch.